branca.go

```diff
@@ -21,128 +34,107 @@ const (
-	if err != nil {
-		return "", errBadKeyLength
+	if !t.IsZero() {
+		timestamp = uint32(t.Unix())
 	}
+	// version[1], timestamp[4], nonce[24], payload[N], auth[16]
+	header := make([]byte, 1+4+24+len(payload)+16)
+	header[0] = version
+	binary.BigEndian.PutUint32(header[1:5], timestamp)
+	copy(header[5:], nonce[:])
+	return b.encrypter.Seal(header[0:1+4+24], nonce[:], payload, header[0:1+4+24])
+}
 
-	ciphertext := xchacha.Seal(nil, nonce, payload, header)
+// EncodeToRaw is like Encode except that it returns the underlying
+// encoded token instead of returning it base62-encoded.
+func (b *Branca) EncodeToRaw(payload []byte) []byte {
+	return b.EncodeToRawAtTime(payload, uuidGen.Next(), time.Now())
+}
 
-	token := append(header, ciphertext...)
-	base62, err := basex.NewEncoding(base62)
-	if err != nil {
-		return "", err
-	}
-	return base62.Encode(token), nil
+// Encode encodes the given payload with the key that b was
+// created with.
+func (b *Branca) Encode(payload []byte) string {
+	return Base62(b.EncodeToRaw(payload))
 }
 
-// DecodeToString decodes the data.
-func (b *Branca) DecodeToString(data string) (string, error) {
-	if len(data) < 62 {
-		return "", errInvalidToken
+// DecodeRaw is like Decode except that it decodes a Branca token that
+// is not base62 encoded.
+func (b *Branca) DecodeRaw(data []byte) ([]byte, time.Time, error) {
+	// version[1], timestamp[4], nonce[24], payload[N], auth[16]
+	if len(data) < 1+4+24+16 {
+		return nil, time.Time{}, errInvalidToken
 	}
-	base62, err := basex.NewEncoding(base62)
+	header := data[0 : 1+4+24]
+	payload, err := b.encrypter.Open(nil, header[1+4:1+4+24], data[1+4+24:], header)
 	if err != nil {
-		return "", errInvalidToken
+		return nil, time.Time{}, err
 	}
-	token, err := base62.Decode(data)
-	if err != nil {
-		return "", errInvalidToken
+	if header[0] != version {
+		return nil, time.Time{}, errInvalidTokenVersion
 	}
-	header := token[0:29]
-	ciphertext := token[29:]
-	tokenversion := header[0]
-	timestamp := binary.BigEndian.Uint32(header[1:5])
-	nonce := header[5:]
-
-	if tokenversion != version {
-		return "", errInvalidTokenVersion
+	timestamp := int64(binary.BigEndian.Uint32(data[1 : 1+4]))
+	var t time.Time
+	if timestamp != 0 {
+		t = time.Unix(timestamp, 0)
 	}
+	return payload, t, nil
+}
 
-	key := bytes.NewBufferString(b.Key).Bytes()
-
-	xchacha, err := chacha20poly1305.NewX(key)
-	if err != nil {
-		return "", errBadKeyLength
-	}
-	payload, err := xchacha.Open(nil, nonce, ciphertext, header)
+// Decode decodes a Branca token. It returns the decrypted payload
+// and the associated timestamp of the token.
+func (b *Branca) Decode(token string) ([]byte, time.Time, error) {
+	data, err := base62Enc.Decode(token)
 	if err != nil {
-		return "", err
-	}
-
-	if b.ttl != 0 {
-		future := int64(timestamp + b.ttl)
-		now := time.Now().Unix()
-		if future < now {
-			return "", errExpiredToken
-		}
+		return nil, time.Time{}, errInvalidToken
```

Comment:
Why not return the same error you receive?

Reply:
Because that was the previous behaviour and it seems somewhat reasonable. We might change it in the future.